markets/storageadapter/ondealsectorcommitted.go

```diff
@@ -20,110 +20,201 @@ type sectorCommittedEventsAPI interface {
 		if err != nil {
-			// TODO: This may be fine for some errors
-			return false, false, xerrors.Errorf("failed to look up deal on chain: %w", err)
+			// Note: the error returned from here will end up being returned
+			// from OnDealSectorPreCommitted so no need to call the callback
+			// with the error
+			return false, false, err
 		}
-		dealID = newDealID
 
-		if sd.State.SectorStartEpoch > 0 {
-			cb(nil)
+		if isActive {
+			// Deal is already active, bail out
+			cb(0, true, nil)
 			return true, false, nil
 		}
 
+		// Not yet active, start matching against incoming messages
 		return false, true, nil
 	}
 
-	var sectorNumber abi.SectorNumber
-	var sectorFound bool
+	// Watch for a pre-commit or prove-commit message to the provider.
+	// It's possible (when the node restarts) that the pre-commit has already
+	// been submitted, in which case we wait for the prove-commit message,
+	// so match both pre-commit and prove-commit messages.
+	matchEvent := func(msg *types.Message) (matched bool, err error) {
+		if msg.To != provider {
+			return false, nil
+		}
 
+		switch msg.Method {
+		case miner.Methods.PreCommitSector:
+			return true, nil
+		case miner.Methods.ProveCommitSector:
+			return true, nil
+		}
+
+		return false, nil
+	}
+
+	// Check if the message params included the deal ID we're looking for.
 	called := func(msg *types.Message, rec *types.MessageReceipt, ts *types.TipSet, curH abi.ChainEpoch) (more bool, err error) {
 		defer func() {
 			if err != nil {
-				cb(xerrors.Errorf("handling applied event: %w", err))
+				cb(0, false, xerrors.Errorf("handling applied event: %w", err))
 			}
 		}()
-		switch msg.Method {
-		case miner.Methods.PreCommitSector:
+
+		// Check if waiting for pre-commit timed out
+		if msg == nil {
+			return false, xerrors.Errorf("timed out waiting for deal %d pre-commit", dealID)
+		}
+
+		var sectorNumber abi.SectorNumber
+		var msgDealIDs []abi.DealID
+		if msg.Method == miner.Methods.PreCommitSector {
+			// If it's a pre-commit message, the deal IDs are in the message parameters
 			var params miner.SectorPreCommitInfo
 			if err := params.UnmarshalCBOR(bytes.NewReader(msg.Params)); err != nil {
 				return false, xerrors.Errorf("unmarshal pre commit: %w", err)
 			}
-
-			dealID, _, err = GetCurrentDealInfo(ctx, ts, api, dealID, proposal, publishCid)
-			if err != nil {
-				return false, err
-			}
-
-			for _, did := range params.DealIDs {
-				if did == dealID {
-					sectorNumber = params.SectorNumber
-					sectorFound = true
-					return true, nil
-				}
-			}
-			return true, nil
-		case miner.Methods.ProveCommitSector:
-			if msg == nil {
-				log.Error("timed out waiting for deal activation... what now?")
-				return false, nil
+			msgDealIDs = params.DealIDs
+			sectorNumber = params.SectorNumber
+		} else {
+			// If it's a prove-commit message, the parameters don't have deal IDs,
+			// just a sector number
+			var params miner.ProveCommitSectorParams
```

Comment:
There is a lot of logic in here to deal with ProveCommitSector messages -- I'm not clear why we're dealing with this. As I see it, the only case of isActive = true we need to deal with is in the called function.

The goal of #4900 is designed to fix the case where a restart happens between pre-commit and prove-commit. We don't need this code for that.

What is the scenario that causes us to miss the pre-commit message entirely? Even if we shut down the node a long time, we will see the pre-commit when we sync the chain.

Reply:
Ah yes that's a good point 👍 

Nice catch @hannahhoward. I think it is worth the effort to refactor the code because it will simplify things significantly. I will do so today.